File: Problems/easy/validParenthesis_STACK.ts

const closingBrackets = [')', '}', ']' ];
const openingBrackets = ['(', '{', '['];
function isPair(start: string | undefined, end: string) {
  if(!start){
    return false;
  }
  return openingBrackets.indexOf(start) !==-1 && openingBrackets.indexOf(start) === closingBrackets.indexOf(end);
}

function isValid(s: string): boolean {
  
  const stack: string[] = [];
  
    for (let item of s) {
      if(stack.length > 0 && closingBrackets.includes(item)) {
        // closing bracket is found
        const lastItem = stack.pop();
        if(isPair(lastItem, item)) {
         
          continue;
        } else {
          return false;
        }
      } else {
        // opening bracket is found
        stack.push(item);
      }
    }
    return stack.length ===0;
}

console.log(isValid('((})){{[}]'));

File: Problems/easy/validParenthesis_STACK.py


#FYI Python implementation

closing_brackets = [')', '}', ']']
opening_brackets = ['(', '{', '[']

def is_pair(start, end):
    if not start:
        return False
    return opening_brackets.index(start) != -1 and opening_brackets.index(start) == closing_brackets.index(end)

def is_valid(s):
    stack = []
    for item in s:
        if stack and item in closing_brackets:
            # closing bracket is found
            last_item = stack.pop()
            if is_pair(last_item, item):
                continue
            else:
                return False
        else:
            # opening bracket is found
            stack.append(item)
    return len(stack) == 0

print(is_valid('((})){{[}]'))
print(is_valid('(({}))'))